### PR TITLE
[LIVY-702]: Submit Spark apps to Kubernetes

### PR DESCRIPTION
**Pull request by jahstreet**
## What changes were proposed in this pull request?

[Jira](https://issues.apache.org/jira/browse/LIVY-702)

This PR is one of the PRs in the series related to the splitting of the base PR https://github.com/apache/incubator-livy/pull/167 to multiple PRs to ease and speed up review and merge processes.

This PR proposes a way to submit Spark apps to Kubernetes cluster. Points covered:
- Submit batch sessions
- Submit interactive sessions
- Monitor sessions, collect logs and diagnostics information
- Restore sessions monitoring after restarts
- GC created Kubernetes resources
- Restrict the set of allowed Kubernetes namespaces

## How was this patch tested?

Unit tests.

Manual testing with Kubernetes on Docker Desktop for Mac v2.1.0.1.
Environment - Helm charts:
- [cluster-base](https://github.com/jahstreet/spark-on-kubernetes-helm/tree/master/charts/cluster-base) with custom-values.yaml:
```yaml
nginx-ingress:
  controller:
    service:
      loadBalancerIP: 127.0.0.1 # my-cluster.example.com IP address (from /etc/hosts)
      loadBalancerSourceRanges: []
cluster-autoscaler:
  enabled: false
oauth2-proxy:
  enabled: false
```

  - [spark-cluster](https://github.com/jahstreet/spark-on-kubernetes-helm/tree/master/charts/spark-cluster) with custom-values.yaml:
```yaml
livy:
  image:
    pullPolicy: Never
    tag: 0.7.0-incubating-spark_2.4.3_2.11-hadoop_3.2.0-dev
  ingress:
    enabled: true
    annotations:
      kubernetes.io/ingress.class: nginx
      kubernetes.io/tls-acme: "true"
      nginx.ingress.kubernetes.io/rewrite-target: /$1
    path: /livy/?(.*)
    hosts:
    - my-cluster.example.com
    tls:
    - secretName: spark-cluster-tls
      hosts:
      - my-cluster.example.com
  persistence:
    enabled: true
  env:
    LIVY_LIVY_UI_BASE1PATH: {value: "/livy"}
    LIVY_SPARK_KUBERNETES_CONTAINER_IMAGE_PULL1POLICY: {value: "Never"}
    LIVY_SPARK_KUBERNETES_CONTAINER_IMAGE: {value: "sasnouskikh/livy-spark:0.7.0-incubating-spark_2.4.3_2.11-hadoop_3.2.0-dev"}
    LIVY_LIVY_SERVER_SESSION_STATE0RETAIN_SEC: {value: "300s"}
    LIVY_LIVY_SERVER_KUBERNETES_ALLOWED1NAMESPACES: {value: "default,test"}
historyserver:
  enabled: false
jupyterhub:
  enabled: true
  ingress:
    enabled: true
    annotations:
      kubernetes.io/ingress.class: nginx
      kubernetes.io/tls-acme: "true"
    hosts:
    - my-cluster.example.com
    pathSuffix: ''
    tls:
    - secretName: spark-cluster-tls
      hosts:
      - my-cluster.example.com
  hub:
    baseUrl: /jupyterhub
    publicURL: "https://my-cluster.example.com"
    activeServerLimit: 10
    # $> openssl rand -hex 32
    cookieSecret: 41b85e5f50222b1542cc3b38a51f4d744864acca5e94eeb78c6e8c19d89eb433
    pdb:
      enabled: true
      minAvailable: 0
  proxy:
    # $> openssl rand -hex 32
    secretToken: cc52356e9a19a50861b22e08c92c40b8ebe617192f77edb355b9bf4b74b055de
    pdb:
      enabled: true
      minAvailable: 0
  cull:
    enabled: false
    timeout: 300
    every: 60
```
- Interactive sessions - Jupyter notebook on JupyterHub with Sparkmagic
- Batch sessions - SparkPi:
```bash
curl -k -H 'Content-Type: application/json' -X POST \
  -d '{
        "name": "SparkPi-01",
        "className": "org.apache.spark.examples.SparkPi",
        "numExecutors": 2,
        "file": "local:///opt/spark/examples/jars/spark-examples_2.11-2.4.3.jar",
        "args": ["10000"],
        "conf": {
            "spark.kubernetes.namespace": "<namespace>"
        }
      }' "https://my-cluster.example.com/livy/batches"
```

### CODE REVIEW

File: server/src/main/scala/org/apache/livy/utils/SparkKubernetesApp.scala [hunk 1]
```diff
@@ -0,0 +1,458 @@
...
+
+  lazy val kubernetesClient: LivyKubernetesClient =
+    KubernetesClientFactory.createKubernetesClient(livyConf)
+
+  private var livyConf: LivyConf = _
+  private var sessionLeakageCheckTimeout: Long = _
+  private var sessionLeakageCheckInterval: Long = _
+  private val leakedAppTags = new java.util.concurrent.ConcurrentHashMap[String, Long]()
+
+  private val leakedAppsGCThread = new Thread() {
```
Comment (yiheng) [inline server/src/main/scala/org/apache/livy/utils/SparkKubernetesApp.scala]: Do we need a GC thread here? RSCDriver will shut down itself if there's no client come in for a while. Please check [this code](https://github.com/apache/incubator-livy/blob/master/rsc/src/main/java/org/apache/livy/rsc/driver/RSCDriver.java#L240)

Comment (jahstreet) [reply server/src/main/scala/org/apache/livy/utils/SparkKubernetesApp.scala]: Actually this GC thread collects `leacked apps`, when Driver cannot be discovered after its submission; look [here](https://github.com/apache/incubator-livy/pull/249/files#diff-9f2a36f41625241cb8e23cf5bcb03751R206-R217).
What about RSCDriver shutdown it comes to play only if Spark Driver has launched, and it works only for Interactive sessions. You may think about Livy GC collecting retained session states... For that could be configured completely different timeouts. Does that make it more clear for you?

File: server/src/main/scala/org/apache/livy/utils/SparkKubernetesApp.scala [hunk 2]
```diff
@@ -0,0 +1,458 @@
...
+      } catch {
+        case _: InterruptedException =>
+          kubernetesDiagnostics = ArrayBuffer("Application stopped by user.")
+          changeState(SparkApp.State.KILLED)
+        case NonFatal(e) =>
+          error("Error while refreshing Kubernetes state", e)
+          kubernetesDiagnostics = ArrayBuffer(e.getMessage)
+          changeState(SparkApp.State.FAILED)
+      } finally {
+        listener.foreach(_.infoChanged(AppInfo(sparkUiUrl = None)))
```
Comment (arunmahadevan) [inline server/src/main/scala/org/apache/livy/utils/SparkKubernetesApp.scala]: How is the user expected to access the driver UI ? Without setting up the ingress and surfacing that URL, it may not be very useful. The original patch handled this and think should be part of the basic requirements.

Comment (jahstreet) [reply server/src/main/scala/org/apache/livy/utils/SparkKubernetesApp.scala]:
```suggestion
        listener.foreach(_.infoChanged(AppInfo()))
```
It shouldn't be unset cause it hasn't been set.

As a first iteration this PR provides a way to submit and track the Spark apps by Livy, as well to integrate Interactive sessions with Notebooks (or whatever). To access Spark UI user still needs to handle access on its own so far. The easiest way I guess is to use `kubectl port-forward ...` manually.

There could be multiple points of view how is it better to compose base PR splitting, unfortunately... But would be nice to let it go at some point. I propose to create small PRs representing single aspect of the whole project each. We can merge them to bigger ones at any time, but splitting out is a bit more painful. In the meantime you can take a look at the following one : https://github.com/apache/incubator-livy/pull/252 .

Or we can roll back to the original PR and just refactor it up to the acceptable state.

WDYT?

**Comment by jahstreet** [comment]
@mgaido91 Could you take a look?

**Comment by jahstreet** [comment]
@mgaido91 ping.

**Comment by mgaido91** [comment]
@jahstreet I am not the best guy to take a look at this honestly. I am reviewing this PR in a few hours, but would be great to have feedbacks also from other people who are more familiar with this part of Livy. cc @vanzin @jerryshao

**Comment by jahstreet** [comment]
> @jahstreet I am not the best guy to take a look at this honestly. I am reviewing this PR in a few hours, but would be great to have feedbacks also from other people who are more familiar with this part of Livy. cc @vanzin @jerryshao

Ah, I see. Will try to ping them. Thanks anyway.

File: server/src/main/scala/org/apache/livy/utils/SparkKubernetesApp.scala [hunk 3]
```diff
@@ -0,0 +1,457 @@
...
+      }
+    }
+
+  override def log(): IndexedSeq[String] =
+    ("stdout: " +: kubernetesAppLog) ++
+      ("\nstderr: " +: (process.map(_.inputLines).getOrElse(ArrayBuffer.empty[String]) ++
+        process.map(_.errorLines).getOrElse(ArrayBuffer.empty[String]))) ++
+      ("\nKubernetes Diagnostics: " +: kubernetesDiagnostics)
+
+  override def kill(): Unit =
```
Comment (mgaido91) [inline server/src/main/scala/org/apache/livy/utils/SparkKubernetesApp.scala]: please add `{` and `}` around methods everywhere, even when not necessary. They help readability.

File: server/src/main/scala/org/apache/livy/utils/SparkKubernetesApp.scala [hunk 4]
```diff
@@ -0,0 +1,457 @@
...
+    pollInterval: Duration,
+    deadline: Deadline): KubernetesApplication = {
+    withRetry(kubernetesClient.getApplications().find(_.getApplicationTag.contains(appTag)))
+    match {
+      case Some(app) => app
+      case None =>
+        if (deadline.isOverdue) {
+          process.foreach(_.destroy())
+          leakedAppTags.put(appTag, System.currentTimeMillis())
+          throw new IllegalStateException(s"No Kubernetes application is found with tag" +
```
Comment (mgaido91) [inline server/src/main/scala/org/apache/livy/utils/SparkKubernetesApp.scala]:
```suggestion
          throw new IllegalStateException("No Kubernetes application is found with tag" +
```

File: server/src/main/scala/org/apache/livy/server/interactive/InteractiveSession.scala [hunk 5]
```diff
@@ -399,7 +399,13 @@ class InteractiveSession(
     app = mockApp.orElse {
       val driverProcess = client.flatMap { c => Option(c.getDriverProcess) }
         .map(new LineBufferedProcess(_, livyConf.getInt(LivyConf.SPARK_LOGS_SIZE)))
-      driverProcess.map { _ => SparkApp.create(appTag, appId, driverProcess, livyConf, Some(this)) }
+      driverProcess.map(_ => SparkApp.create(appTag, appId, driverProcess, livyConf, Some(this)))
+        .orElse {
```
[the remark this answers is not in the document]
Comment (jahstreet) [reply server/src/main/scala/org/apache/livy/server/interactive/InteractiveSession.scala]: sorry, i haven't got the idea. what is `nit` ?

File: server/src/main/scala/org/apache/livy/utils/SparkKubernetesApp.scala [hunk 6]
```diff
@@ -0,0 +1,457 @@
...
+
+  private[utils] object KubernetesApplicationState {
+    val PENDING = "pending"
+    val RUNNING = "running"
+    val SUCCEEDED = "succeeded"
+    val FAILED = "failed"
+  }
+
+  private[utils] def mapKubernetesState(
+    kubernetesAppState: String,
```
[the remark this answers is not in the document]
Comment (jahstreet) [reply server/src/main/scala/org/apache/livy/utils/SparkKubernetesApp.scala]: an here. could you please explain what do you mean?

[the remark this answers is not in the document]
Comment (mgaido91) [reply server/src/main/scala/org/apache/livy/utils/SparkKubernetesApp.scala]: add other 2 spaces indent here, like:

```suggestion
      kubernetesAppState: String,
```

File: server/src/main/scala/org/apache/livy/server/interactive/InteractiveSession.scala (hunk 5 above)
[the remark this answers is not in the document]
Comment (mgaido91) [reply server/src/main/scala/org/apache/livy/server/interactive/InteractiveSession.scala]: it is the short for `nitpicking`.

I meant to put `.oElse {` immediately after the `(` on the line above. It is just a style thing.

[the remark this answers is not in the document]
Comment (jahstreet) [reply server/src/main/scala/org/apache/livy/server/interactive/InteractiveSession.scala]: Exceeds line length then

File: server/src/main/scala/org/apache/livy/utils/SparkKubernetesApp.scala [hunk 7]
```diff
@@ -0,0 +1,457 @@
...
+          // kill the app if found it or remove it if exceeding a threshold
+          val leakedApps = leakedAppTags.entrySet().iterator()
+          val now = System.currentTimeMillis()
+          val apps = withRetry(kubernetesClient.getApplications()).groupBy(_.getApplicationTag)
+          while (leakedApps.hasNext) {
+            val leakedApp = leakedApps.next()
+            apps.get(leakedApp.getKey) match {
+              case Some(seq) =>
+                seq.foreach(app =>
+                  if (withRetry(kubernetesClient.killApplication(app))) {
```
[the remark this answers is not in the document]
Comment (jahstreet) [reply server/src/main/scala/org/apache/livy/utils/SparkKubernetesApp.scala]: Nice catch!

File: server/src/main/scala/org/apache/livy/utils/SparkKubernetesApp.scala [hunk 8]
```diff
@@ -0,0 +1,457 @@
...
+  private val leakedAppTags = new java.util.concurrent.ConcurrentHashMap[String, Long]()
+
+  private val leakedAppsGCThread = new Thread() {
+    override def run(): Unit = {
+      while (true) {
+        if (!leakedAppTags.isEmpty) {
+          // kill the app if found it or remove it if exceeding a threshold
+          val leakedApps = leakedAppTags.entrySet().iterator()
+          val now = System.currentTimeMillis()
+          val apps = withRetry(kubernetesClient.getApplications()).groupBy(_.getApplicationTag)
```
[the remark this answers is not in the document]
Comment (jahstreet) [reply server/src/main/scala/org/apache/livy/utils/SparkKubernetesApp.scala]: Nice catch!

File: server/src/main/scala/org/apache/livy/server/interactive/InteractiveSession.scala (hunk 5 above)
[the remark this answers is not in the document]
Comment (mgaido91) [reply server/src/main/scala/org/apache/livy/server/interactive/InteractiveSession.scala]:
```
driverProcess.map(
  _ => SparkApp.create(appTag, appId, driverProcess, livyConf, Some(this))).orElse {
```

[the remark this answers is not in the document]
Comment (jahstreet) [reply server/src/main/scala/org/apache/livy/server/interactive/InteractiveSession.scala]: Ahh, ok!

**Comment by jahstreet** [comment]
Build vailure due to Travis:
```
No output has been received in the last 10m0s, this potentially indicates a stalled build or something wrong with the build itself.
Check the details on how to adjust your build configuration on: https://docs.travis-ci.com/user/common-build-problems/#build-times-out-because-no-output-was-received
```
@jerryshao @vanzin could you take a look please?

**Comment by jahstreet** [comment]
@yiheng @arunmahadevan @mgaido91 
Anything else from your side?

**Comment by jahstreet** [comment]
Rebased

**Comment by jahstreet** [comment]
Rebased to `master`.

File: server/src/main/scala/org/apache/livy/server/interactive/InteractiveSession.scala [hunk 9]
```diff
@@ -402,6 +402,9 @@ class InteractiveSession(
 
       if (livyConf.isRunningOnYarn() || driverProcess.isDefined) {
         Some(SparkApp.create(appTag, appId, driverProcess, livyConf, Some(this)))
+      } else if (livyConf.isRunningOnKubernetes()) {
+        // Create SparkKubernetesApp anyway to recover app monitoring on Livy server restart
+        Some(SparkApp.create(appTag, appId, driverProcess, livyConf, Some(this)))
```
Comment (ajbozarth) [inline server/src/main/scala/org/apache/livy/server/interactive/InteractiveSession.scala]: If this is just the same line as 404 why is it in it's own `else if` block? Wouldn't it make more sense to add ` || livyConf.isRunningOnKubernetes()` to  the if on line 403?

Comment (jahstreet) [reply server/src/main/scala/org/apache/livy/server/interactive/InteractiveSession.scala]: Ahh, nice catch, agree.

**Comment by lukatera** [comment]
can we merge this? :)

**Comment by jahstreet** [comment]
> can we merge this? :)

I would also love to! Opened for the suggestions on how to get closer to it.

**Comment by SarnathK** [comment]
Is there a timeline when this will get integrated with Livy? This would help us run Jupyter on Spark on Kubernetes. Any ETA will be very helpful! Thanks!

**Comment by jahstreet** [comment]
Hi @SarnathK , I've tried to contact the community multiple times via mailing lists with no luck to push this forward. 
I'm tracking the activity around this work and have a list of patches on top of it in the backlog. Also I'm always ready to provide the full support around on up Livy on Kubernetes.

**Comment by ajbozarth** [comment]
@jerryshao do you have bandwidth to review this, I've done a partial review above, but need another pair of eyes.

**Comment by jerryshao** [comment]
I can take a chance to review this, but I'm not an export of k8s, may not fully understand the pros and cons of the implementation.

**Comment by jahstreet** [comment]
@jerryshao I check all the notifications on this PR, feel free to add comments on uncertainties, will be glad to describe the idea.

**Comment by SarnathK** [comment]
@jahstreet sure. Please feel free to add me to any relevant thread. I am fully convinced that this is a great solution for many organizations that deal with high volume data. Many are looking for but don't know what and how to do. I really hope this one gets integrated with Livy. Many thanks again @jahstreet for the contribution!

**Comment by jahstreet** [comment]
Rebased to master.

**Comment by cyliu0204** [comment]
We use spark magic in jupyter to connect livy to start the Spark cluster in kubernetes. We have started to use this patch in our work. As more and more applications migrate towards the cloud, this patch is definitely very valuable.  hopefully this feature can be merged asap，so that more people can see and use this feature

**Comment by jahstreet** [comment]
> I can take a chance to review this, but I'm not an export of k8s, may not fully understand the pros and cons of the implementation.

@jerryshao , have you given it a try already?

**Comment by lukatera** [comment]
I had some issues with using ZooKeeper as a recovery method. When one of the zookeepers is not up in Kubernetes, its address is not published (i.e. there is none), and thus Livy cannot determine the zookeeper host. With the current version of ZooKeeper, Livy crashes because of a bug in ZooKeeper client. This has been fixed though in more recent versions (3.5.2+): [ZOOKEEPER-1576](https://zookeeper.apache.org/doc/r3.5.2-alpha/releasenotes.html).

In short, Livy cannot start if 2 out of 3 ZooKeepers are up at the moment, even though it should be able to do so. Bumping the ZooKeeper client version should resolve the issue.

@jahstreet Can we update ZooKeeper version in this PR as well?

**Comment by jahstreet** [comment]
Hi @lukatera , upgrading ZooKeeper client is out of scope of this PR and is a good candidate for a separate one. Feel free to open it. Later we can rebase this one to master once it is merged and you can cherry-pick it locally (or to your fork) to add it to this PR code. Will it work for you?

**Comment by lukatera** [comment]
> Hi @lukatera , upgrading ZooKeeper client is out of scope of this PR and is a good candidate for a separate one. Feel free to open it. Later we can rebase this one to master once it is merged and you can cherry-pick it locally (or to your fork) to add it to this PR code. Will it work for you?

Sure, makes sense. Thanks :)
Is there a single branch somewhere where all the K8S functionality is present so I can try to build it locally?

**Comment by jahstreet** [comment]
@lukatera , you can check that PR: https://github.com/apache/incubator-livy/pull/167 . Basically this PR is the clean backport from the former one.

**Comment by gmcoringa** [comment]
Thanks @jahstreet for your effort. 
Tested with:
* kubernetes 1.15.11
* Spark 2.4.5

What is holding this pr and https://github.com/apache/incubator-livy/pull/252?

Also, @jahstreet if you could add me in the threads I can explain our use case, where we use jupyter to schedule spark jobs on kubernetes.

**Comment by jahstreet** [comment]
@gmcoringa , thank you for the feedback. Can I have your e-mail to add you to the thread?

**Comment by gmcoringa** [comment]
> @gmcoringa , thank you for the feedback. Can I have your e-mail to add you to the thread?
santosfabianov@gmail.com

**Comment by jesinity** [comment]
is there any hope to see this merged anytime soonish?

File: rsc/src/main/java/org/apache/livy/rsc/driver/RSCDriver.java [hunk 10]
```diff
@@ -169,6 +169,13 @@ private void initializeServer() throws Exception {
     // on the cluster, it would be tricky to solve that problem in a generic way.
     livyConf.set(RPC_SERVER_ADDRESS, null);
 
+    // If we are running on Kubernetes, set RPC_SERVER_ADDRESS from "spark.driver.host" option,
+    // which is set in class org.apache.spark.deploy.k8s.features.DriverServiceFeatureStep:
+    // line 61: val driverHostname = s"$resolvedServiceName.${kubernetesConf.namespace()}.svc"
+    if (livyConf.isRunningOnKubernetes()) {
+      livyConf.set(RPC_SERVER_ADDRESS, conf.get("spark.driver.host"));
+    }
+
```
Comment (idzikovsky) [inline rsc/src/main/java/org/apache/livy/rsc/driver/RSCDriver.java:178]: By some reason this version does not work for me like it was with the same piece from https://github.com/apache/incubator-livy/pull/167.

I got the same exception as here: https://github.com/apache/incubator-livy/pull/167#issuecomment-485878996

I'm not sure why, but it seems like it's because the `livy.spark.master` property is not set inside of Spark Driver of Livy Session on my env.
While the same piece from https://github.com/apache/incubator-livy/pull/167 works fine:
```
    if (conf.get("spark.master").startsWith("k8s")) {
      livyConf.set(RPC_SERVER_ADDRESS, conf.get("spark.driver.host"));
    }
```

But I'm not using code from your branch, but rather backporting your patch to our Livy build which is based on older Livy version, so maybe it's the cause.

On the other hand, during quick lookup I've not found any code that bypass `livy.spark.master` property into driver, so I see no reason why it should work here.
If I get a chance to test this issue on build from this PR, I'll share what get.

Comment (jahstreet) [reply rsc/src/main/java/org/apache/livy/rsc/driver/RSCDriver.java:178]: Might be related to the backporting indeed. Will be happy to help once you get more debug info.

**Comment by jahstreet** [comment]
Good news, yesterday I've upgraded the helm charts to Spark 3.0.1 which unlocks K8s API 1.18.x usage. Feel free to try it out with [this guide](https://github.com/JahstreetOrg/spark-on-kubernetes-helm/blob/master/DOCUMENTATION.md).
In the meantime I'm going to backport the required changes to this PR.

**Comment by kyprifog** [comment]
FYI I recieved:  
```
Error: chart requires kubeVersion: 1.11.0 - 1.18.0 which is incompatible with Kubernetes v1.18.9
```
Did you mean `1.18.x` or `1.18.0`?

**Comment by jahstreet** [comment]
Tested with 1.18.0 only, so yeah, 1.18.0 would be better to say. Though I haven't tried with higher versions, in case you can modify the chart locally and remove the kubeVersion restrictions you can try it out on 1.18.9.

**Comment by kyprifog** [comment]
After modifying the limit bounds on chart locally I just tested and it worked!

image
image (1)

**Comment by jahstreet** [comment]
> After modifying the limit bounds on chart locally I just tested and it worked!

Thx for the test, I'll patch the chart then.

**Comment by kyprifog** [comment]
Sure thing, here you go:  https://github.com/JahstreetOrg/spark-on-kubernetes-helm/pull/43/files

**Comment by zhenlohuang** [comment]
It is a very nice feature! Thanks for your efforts. 

@ Committees  Do we have a plan to merge this feature?

File: rsc/src/main/java/org/apache/livy/rsc/driver/RSCDriver.java (hunk 10 above)
Comment (idzikovsky) [reply rsc/src/main/java/org/apache/livy/rsc/driver/RSCDriver.java:178]: I've tried to use the latest Livy with your patches and the issue has not appeared. So it seems like I got that problem because I've backported something wrong.
Anyway, thank you for help!

**Comment by butaozhang** [comment]
Good feature! Why didn't the livy community merge this feature?  Livy community seems very inactive and no code updates since  the last year. . .

**Comment by PerilousApricot** [comment]
This looks very useful! Any plans to merge?

**Comment by jahstreet** [comment]
@PerilousApricot thx for the feedback. FYI I do keep an eye on it waiting for the reviewers and at least 1 approval from the repo admin so that someone could press the merge button in the end. Your review and (hopefully) approval would help a lot 🙏 . Also the short story about how you and/or your team would benefit from this feature might be helpful to attract more attention to the PR.

**Comment by PerilousApricot** [comment]
@jahstreet I've only done some preliminary messing around with your branch, but it works on k8s for me.

I'm in academia, which means two things -- we have on-prem deployments, and we have chaotic use patterns. We have 200 users trying to use 2k cores via notebooks. Traditional spark deployments are inefficient because we have to statically partition the cluster, which means that a) cores stay idle b) there is no shared caches between user applications. 

Really livy is the only good solution for this form of multitenancy, but it doesn't support k8s in master, so it's not good

**Comment by haripriya-b** [comment]
Hi @jahstreet , are there any plans for adding support for Spark 3.2 in this PR?

**Comment by jahstreet** [comment]
Hi @haripriya-b ,

Eventually yes. Going to get to it in the coming months to sync with the latest updates in the world.

**Comment by nanaones** [comment]
> 

also very excited about the release of this feature. 👍

**Comment by Rustem** [comment]
Really great work @jahstreet . Any plans to upgrade with most recent kuber version ?

**Comment by jahstreet** [comment]
Hi @Rustem , thanks for your feedback. Hope to get to it in the coming month or 2.

**Comment by alexlang74** [comment]
I agree - this is excellent work, @jahstreet ! Too bad that the Livy community is dormant, and not able to pick up your changes... I echo @Rustem 's request: I could easily live with Spark 3.0 (or even "backlevel" Spark 2.4), but would be great to be able to support a newer Kubernetes version (e.g., 1.23)...

**Comment by nanaones** [comment]
Why is this important feature PR still open?
Any problem?
@jahstreet

**Comment by jdcohen220** [comment]
We would love to start using this once its released.

**Comment by ringtail** [comment]
Hey guys, are there still block issues left?

**Comment by anistal** [comment]
I agree with all the comments: a lot of effort that should be at least considered

File: rsc/src/main/java/org/apache/livy/rsc/RSCConf.java [hunk 11]
```diff
@@ -151,6 +152,12 @@ public String findLocalAddress() throws IOException {
     return address.getCanonicalHostName();
   }
 
+  public boolean isRunningOnKubernetes() {
+    return Optional.ofNullable(get("livy.spark.master"))
+            .filter(s -> s.startsWith("k8s"))
+            .isPresent();
+  }
+
```
Comment (jinhaichen) [inline rsc/src/main/java/org/apache/livy/rsc/RSCConf.java:160]: this function may always return false, since "livy.spark.master" will not get by RSCConf

Comment (idzikovsky) [reply rsc/src/main/java/org/apache/livy/rsc/RSCConf.java:160]: Yeah, seems like that's what I've faced in https://github.com/apache/incubator-livy/pull/249#discussion_r493348458

However it seems like it's not affecting functionality, as this function is used while setting `RPC_SERVER_ADDRESS` here:
https://github.com/apache/incubator-livy/pull/249/files/b87c0cebb65ce7f34e6b4b6b738095be6254cf69#diff-43114318c4b009c2404f7eb326a84c184fb1501a3237c49a771df851d0f6f328R172-R178

And the value of `RPC_SERVER_ADDRESS` is not used anyway since Livy 0.7 because of things I've explained in #388.

**Comment by askhatri** [comment]
I have validated this fix in [the new git branch](https://github.com/askhatri/incubator-livy/tree/validating-livy-702-fix). I found that the fix is working as expected. The detailed steps used during the validation are documented at [README.md](https://github.com/askhatri/incubator-livy/tree/validating-livy-702-fix/dev/kubernetes). Should we consider merging the fix?

**Comment by lmccay** [comment]
@askhatri - I would say that you can merge this given your review and testing.

**Comment by gyogal** [comment]
Thank you @lmccay! This LGTM as well and based on @askhatri 's testing, if there are no objections, I think we should go ahead and merge this. We can address the remaining issues in separate tickets.

**Comment by jahstreet** [comment]
Happy to see things going in this PR 🎉 . Thank you a lot folks for putting the effort in reviewing and testing the change. 

@askhatri can I help you with anything to test the work?

I'd love to pay my time to finalize the activity here and then I have something to offer on top of it to contribute with.

**Comment by askhatri** [comment]
> Happy to see things going in this PR 🎉 . Thank you a lot folks for putting the effort in reviewing and testing the change.
> 
> @askhatri can I help you with anything to test the work?
> 
> I'd love to pay my time to finalize the activity here and then I have something to offer on top of it to contribute with.

Thank you, @jahstreet for offering your help in testing.

During my initial testing, I found that the code is working as expected. Only one observation is that we might need to upgrade [spark-on-kubernetes-helm](https://github.com/JahstreetOrg/spark-on-kubernetes-helm) to support Helm Chart 3.x and Kubernetes latest version 1.24 or higher.

**Comment by jahstreet** [comment]
> > Happy to see things going in this PR 🎉 . Thank you a lot folks for putting the effort in reviewing and testing the change.
> > @askhatri can I help you with anything to test the work?
> > I'd love to pay my time to finalize the activity here and then I have something to offer on top of it to contribute with.
> 
> Thank you, @jahstreet for offering your help in testing.
> 
> During my initial testing, I found that the code is working as expected. Only one observation is that we might need to upgrade [spark-on-kubernetes-helm](https://github.com/JahstreetOrg/spark-on-kubernetes-helm) to support Helm Chart 3.x and Kubernetes latest version 1.24 or higher.

Looking into it, not 100% sure we can get latest K8s version, given latest Spark 3.3.2 works with Fabric8 client 5.12.2 which aims for K8s <= 1.23.13 (as per [compatibility matrix](https://github.com/fabric8io/kubernetes-client#compatibility-matrix)). Will share my findings ...

**Comment by askhatri** [comment]
> > > Happy to see things going in this PR 🎉 . Thank you a lot folks for putting the effort in reviewing and testing the change.
> > > @askhatri can I help you with anything to test the work?
> > > I'd love to pay my time to finalize the activity here and then I have something to offer on top of it to contribute with.
> > 
> > 
> > Thank you, @jahstreet for offering your help in testing.
> > During my initial testing, I found that the code is working as expected. Only one observation is that we might need to upgrade [spark-on-kubernetes-helm](https://github.com/JahstreetOrg/spark-on-kubernetes-helm) to support Helm Chart 3.x and Kubernetes latest version 1.24 or higher.
> 
> Looking into it, not 100% sure we can get latest K8s version, given latest Spark 3.3.2 works with Fabric8 client 5.12.2 which aims for K8s <= 1.23.13 (as per [compatibility matrix](https://github.com/fabric8io/kubernetes-client#compatibility-matrix)). Will share my findings ...

Thank you @jahstreet

**Comment by satishdalli** [comment]
Absolutely great work @jahstreet. We are waiting for PR with Helm 3 and k8s latest versions support. currently, we are using livy with consistent resources. we want to try this in AKS, EKS, and on-prem k8s clusters as a server-less livy.

**Comment by lmccay** [comment]
Hey Folks - I noticed this JIRA is listed for 0.8.0 and would like to get a sense for how far out this may be.
I'm trying to burn down the release blockers so that we can tackle the process of the first release since the reboot.
If I don't hear otherwise, I will move it to 0.9.0 and we can discuss pulling it back in on the JIRA - if need be.

**Comment by jahstreet** [comment]
Hi @lmccay , thank you for keeping eye on it.

I think this PR is already battle tested and good to go. There is already a chain of work done on top of it by me and other people who left feedback on this chunk of work. If we make it a part of master I'm 100% sure I won't be the only one pushing Livy project to the world of K8s. Besides that, seeing the progress after the years of waiting would boost the motivation to continue contributions... So, including the upgrade of dependencies to support latest K8s and Spark versions, I propose to tackle those in separate PRs. How do you feel about merging it to 0.8 and is there anything formal we should do to make it happen?

(resolving the merge conflicts in the meantime)

**Comment by idzikovsky** [comment]
Yeah. We've been using Livy on K8s with fixes from this PR and PR #252 for near 2 years now, in different configurations. Including different Spark versions starting from 2.4.4 and up to 3.3.1 (however some fixes where made in Livy itself to make it compatible with Spark 3.3).
Everything seems to be fine by now, except for the setup with Istio.
To resolve those issues, we applied fix from #388 and reverted the thing which I described in https://github.com/apache/incubator-livy/pull/249#discussion_r1137058321 and https://github.com/apache/incubator-livy/pull/249#discussion_r493348458.

But in general everything works fantastic.
Thank you @jahstreet

**Comment by lmccay** [comment]
@jahstreet and @idzikovsky  - if we merge this before branching from 0.8.0 and it doesn't cause any issues due to other work not being there yet then I would have no problem with doing so. I'm personally not in a position to +1 the merge. Would someone that has tested it and/or tested with it inplace but not necessarily exercised be able to +1 it as a review?

**Comment by jahstreet** [comment]
@lmccay I think we need to summon a maintainer with K8s expertise, is that something you expect? Do you know any name(s) we can put here and follow-up on that together?

**Comment by lmouhib** [comment]
@jahstreet The implementation is great. I got couple of remarks/questions on the way it handle the logs and the ingress, the current implementation is opinionated for ngnix and lokki, while both are widely adopted, how can someone with another ingress controller use their own? I understand the log might be a bit more challenging to provide the native UI integration, in that case maybe offer the possibility to use a sidecar for log shipping?

**Comment by ozsoyler** [comment]
Hi. Is that possible to merge this into apache-incubator-livy officially on new version "0.9.0"? I guess if it merges, we will be able to use livy on kubernetes for launching our awesome spark jobs! Thanks..

**Comment by askhatri** [comment]
Adding Kubernetes support to Apache Livy is a valuable enhancement for the next release. Should we consider merging this into the master branch?

**Comment by lmccay** [comment]
@askhatri - yes, I think we should push for an active approval on the merge of this. We can likely mark this as an important contribution for the next release. Let's move this forward now.

**Comment by jesinity** [comment]
I've been following this issue since forever,  as it would have been highly beneficial somewhat like 2 companies ago.
The work has been done but it was never merged. Is there maybe some passive obstruction against it? Looks like so.

**Comment by idzikovsky** [comment]
@jesinity there is no passive obstruction against it here.
The problem is that there are no active Livy maintainers who are willing to review this and/or familiar enough with K8s.

**Comment by devstein** [comment]
To help with the K8s review/perspective, I know there are teams that have been using this branch in production for years (cc @jpugliesi).

**Comment by askhatri** [comment]
I have created a new https://github.com/apache/incubator-livy/pull/451 with the update includes a newer version of the Kubernetes client and adds code to display the Spark UI. CC: @jahstreet
